src/Processors/QueryPlan/ReadFromMemoryStorageStep.cpp

```diff
@@ -93,29 +94,52 @@ class MemorySource : public ISource
     InitializerFunc initializer_func;
 };
 
-ReadFromMemoryStorageStep::ReadFromMemoryStorageStep(Pipe pipe_) :
-    SourceStepWithFilter(DataStream{.header = pipe_.getHeader()}),
-    pipe(std::move(pipe_))
+ReadFromMemoryStorageStep::ReadFromMemoryStorageStep(QueryPlan & query_plan_,
+                                                     const SelectQueryInfo & query_info_,
+                                                     ContextPtr context_,
+                                                     const Names & columns_to_read_,
+                                                     const StorageSnapshotPtr & storage_snapshot_,
+                                                     const size_t num_streams_,
+                                                     const bool delay_read_for_global_sub_queries_) :
+    SourceStepWithFilter(DataStream{}),
```

Comment:
We can't set an empty header here. It should be known after construction. Otherwise, some plan-level optimizations will not work.

Use: `storage_snapshot_->getSampleBlockForColumns(column_names)`. It's the sam way used in MemorySource.

Reply:
thx for getSampleBlockForColumns